parquet/src/main/java/org/apache/iceberg/parquet/ParquetFileMerger.java

```diff
@@ -0,0 +1,643 @@
+ *   ParquetFileMerger.mergeFiles(
+ *       inputFiles, encryptedOutputFile, result.schema(), firstRowIds,
+ *       rowGroupSize, columnIndexTruncateLength, result.metadata());
+ * }
+ * </pre>
+ */
+public class ParquetFileMerger {
+  // Default buffer sizes for DeltaBinaryPackingValuesWriter
+  private static final int DEFAULT_INITIAL_BUFFER_SIZE = 64 * 1024; // 64KB
+  private static final int DEFAULT_PAGE_SIZE_FOR_ENCODING = 64 * 1024; // 64KB
+
+  private ParquetFileMerger() {
+    // Utility class - prevent instantiation
+  }
+
+  /**
+   * Validates that all input files can be merged.
+   *
+   * <p>This method validates that:
+   *
+   * <ul>
+   *   <li>All files are valid Parquet format (detected by reading Parquet footer)
+   *   <li>All files have identical schemas
+   *   <li>Files are not encrypted (detected by ParquetCryptoRuntimeException)
+   *   <li>If physical row lineage columns exist, all values are non-null
+   * </ul>
+   *
+   * <p>This method works with any Iceberg FileIO implementation (S3FileIO, GCSFileIO, etc.).
+   *
+   * @param inputFiles List of Iceberg input files to validate
+   * @return true if files can be merged, false otherwise
+   */
+  public static boolean canMerge(List<InputFile> inputFiles) {
+    try {
+      if (inputFiles == null || inputFiles.isEmpty()) {
+        return false;
+      }
+
+      // Read schema from the first file
+      MessageType firstSchema = readSchema(inputFiles.get(0));
+
+      // Check if schema has physical row lineage columns
+      boolean hasRowIdColumn = firstSchema.containsField(MetadataColumns.ROW_ID.name());
+      boolean hasSeqNumColumn =
+          firstSchema.containsField(MetadataColumns.LAST_UPDATED_SEQUENCE_NUMBER.name());
+
+      // Validate all files have the same schema
+      for (int i = 1; i < inputFiles.size(); i++) {
+        MessageType currentSchema = readSchema(inputFiles.get(i));
+
+        if (!firstSchema.equals(currentSchema)) {
+          return false;
+        }
+      }
+
+      // If there are physical row lineage columns, validate no nulls
+      if (hasRowIdColumn || hasSeqNumColumn) {
+        validateRowLineageColumnsHaveNoNulls(inputFiles);
+      }
+
+      return true;
+    } catch (RuntimeException | IOException e) {
+      // Returns false for:
+      // - Non-Parquet files (IOException when reading Parquet footer)
+      // - Encrypted files (ParquetCryptoRuntimeException extends RuntimeException)
+      // - Files with null row lineage values (IllegalArgumentException from
+      // validateRowLineageColumnsHaveNoNulls)
+      // - Any other validation failures
+      return false;
+    }
+  }
+
+  /**
+   * Reads the Parquet schema from an Iceberg InputFile.
+   *
+   * @param inputFile Iceberg input file to read schema from
+   * @return MessageType schema of the Parquet file
+   * @throws IOException if reading fails
+   */
+  public static MessageType readSchema(InputFile inputFile) throws IOException {
```

Comment:
If this is public only because of the tests, please find another way

Reply:
The two methods: readSchema () and readMetadata () are public because they are used in production code in the Spark module. 
File: spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/SparkParquetFileMergeRunner.java:394-395
  MessageType schema = ParquetFileMerger.readSchema(inputFiles.get(0));
  Map<String, String> metadata = ParquetFileMerger.readMetadata(inputFiles.get(0));